KEP-3158: Optional maximum etcd page size on every list requests (#3252)
kubernetes/enhancements: keps/sig-api-machinery/3158-optional-maximum-limit-on-list-requests-to-etcd/README.md

easier for reviewers to cite specific portions, and to minimize diff churn on
updates.

[documentation style guide]: https://github.com/kubernetes/community/blob/master/contributors/guide/style-guide.md
-->
Performing API queries that return all of the objects of a given resource type (GET /api/v1/pods, GET /api/v1/secrets) without pagination can lead to significant variations in peak memory use on etcd. 

This proposal covers an approach where the apiserver makes multiple paginated list calls to etcd instead of a single large unpaginated call to reduce the peak memory usage of etcd, thereby reducing its chances of Out-of-memory crashes. In setups where etcd is colocated with apiserver on the same instance, it can buy more memory for apiserver. 

## Motivation

<!--
This section is for explicitly listing the motivation, goals, and non-goals of
this KEP.  Describe why the change is important and the benefits to users. The
motivation section can optionally provide links to [experience reports] to
demonstrate the interest in a KEP within the wider Kubernetes community.

[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
-->

> Protect etcdfrom out of memory crashing and prevent its cascading unpredicable failures. APIServer and other downstream components recovering like re-creating gRPC connections to etcd, rebuilding all caches and resuming internal reconciler can be expensive. Lower etcd RAM consumption at peak time to allow auto scaler to increase memory budget in time.

Reviewer: ```suggestion
Protect etcd from out of memory crashing and prevent its cascading unpredictable failures. kube-apiserver and other downstream components recovering like re-creating gRPC connections to etcd, rebuilding all caches and resuming internal reconciler can be expensive. Lower etcd RAM consumption at peak time to allow autoscalers to increase their memory budget in time.
```

> In Kubernetes, etcd only has a single client: APIServer. It is feasible to control how requests are sent from APIServer to etcd.

Reviewer: ```suggestion
In Kubernetes, etcd only has a single client: kube-apiserver. It is feasible to control how requests are sent from kube-apiserver to etcd.
```

### Goals

<!--
List the specific goals of the KEP. What is it trying to achieve? How will we
know that this has succeeded?
-->

- Reduce etcd memory consumption when kube clients list a large number of resources

### Non-Goals

<!--
What is out of scope for this KEP? Listing non-goals helps to focus discussion
and make progress.
-->

- Reduce apiserver memory usage for unpaginated list calls
- Implement etcd server side streaming serves apiserver List
- Implement etcd QoS
- Reduce list call load on etcd using priority & fairness settings on the apiserver

## Proposal

<!--
This is where we get down to the specifics of what the proposal actually is.
This should have enough detail that reviewers can understand exactly what
you're proposing, but should not include things like API designs or
implementation. What is the desired outcome and how do we measure success?.
The "Design Details" section below is for the real
nitty-gritty.
-->

### Desired outcome
By default, this proposal is a no-op. 

> If this feature is opted-in and `max-list-etcd-limit` flag on APIServer is set to `x` where `x` >= 500

Reviewer: ```suggestion
If the relevant feature gate is enabled and the `max-list-etcd-limit` command line argument to kube-apiserver is set to `x`, where `x >= 500`, then:
```

> - KubeAPIServer splits requests to etcd into multiple pages. The max etcd page size is `x`. If user provided the limit `y` is smaller or equal to `x`, those requests are intact.

Reviewer: ```suggestion
- kube-apiserver splits requests to etcd into multiple pages. The max etcd page size is `x`. If user provided the limit `y` is smaller or equal to `x`, those requests are intact.
```

- The returned response is identical to the one without split. 

### How do we measure success
If this feature is opted-in and `max-list-etcd-limit` flag on APIServer is set to `x` where `x` >= 500

- The number of etcd OOM-killed incidents is reduced. 
- etcd memory consumption is less than before. 
- Only a small set of user list requests are split by APIServer. 
- For impacted user list requests, added latency about apiserver_list_from_storage_duration_p50/p90/p99 is within 5s
- apiserver_list_duration_p99 is within defined scalability SLO

### User Stories (Optional)

<!--
Detail the things that people will be able to do if this KEP is implemented.
Include as much detail as possible so that people can understand the "how" of
the system. The goal here is to make this feel real for users without getting
bogged down.
-->

#### Story 1
A user deployed an application as daemonset that queries pods belonging to the node it is running. It translates to range query etcd from `/registry/pods/` to `/registry/pods0` and each such query payload is close to 2GiB. 

> ```

Reviewer: nit, you could mark this as JSON.

<!--
**Note:** When your KEP is complete, all of these comment blocks should be removed.

> To get started with this template:

Reviewer 2: Can we remove these comment blocks to make reviewing easier? Given you've already filled the details.

-->

[kubernetes.io]: https://kubernetes.io/
[kubernetes/enhancements]: https://git.k8s.io/enhancements
[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
[kubernetes/website]: https://git.k8s.io/website

## Summary

<!--
This section is incredibly important for producing high-quality, user-focused
documentation such as release notes or a development roadmap. It should be
possible to collect this information before implementation begins, in order to
avoid requiring implementors to split their attention between writing release
notes and implementing the feature itself. KEP editors and SIG Docs
should help to ensure that the tone and content of the `Summary` section is
useful for a wide audience.

A good summary is probably at least a paragraph in length.

Both in this section and below, follow the guidelines of the [documentation
style guide]. In particular, wrap lines to a reasonable length, to make it
easier for reviewers to cite specific portions, and to minimize diff churn on
updates.

[documentation style guide]: https://github.com/kubernetes/community/blob/master/contributors/guide/style-guide.md
-->

> Performing API queries that return all of the objects of a given resource type (GET /api/v1/pods, GET /api/v1/secrets) without pagination can lead to significant variations in peak memory use on etcd and APIServer. 

Reviewer 2: Apiserver memory usage is also mentioned here, but this proposal is primarily helping optimize etcd memory only. Worth making this clear and add "reducing apiserver memory usage for unpaginated list calls" as a non-goal.

Reviewer 2: One thing to note though is this may indirectly also help apiserver in setups where etcd is colocated with apiserver on the same instance. So reducing etcd mem usage, can buy more memory for apiserver.

Author: > One thing to note though is this may indirectly also help apiserver in setups where etcd is colocated with apiserver on the same instance. So reducing etcd mem usage, can buy more memory for apiserver. 

That's a real good point for other cloud-providers to consider use this feature.

> This proposal covers an approach that splitting full resource list to protect etcd from out of memory crashing

Reviewer 2: "an approach where the apiserver makes multiple paginated list calls to etcd instead of a single large unpaginated call to reduce the peak memory usage of etcd, thereby reducing its chances of Out-of-memory crashes"

^ would be clearer

In Kubernetes, etcd only has a single client: APIServer. It is feasible to control how requests are sent from APIServer to etcd.

[APIListChunking](https://kubernetes.io/docs/reference/using-api/api-concepts/#retrieving-large-results-sets-in-chunks) should be in effect in most APIServer set up, `client-go/tools/pager` and `client-go/informers` are good libraries to use and they follows the pagination rule of thumb to reduce server load. 

However, there are still some native clients calling server without any limit to get a full list of resource. In addition, etcd buffers those request responses entirely in memory before sending the full set. It does not [stream the result](https://github.com/etcd-io/etcd/pull/12343#issuecomment-784008186) gracefully. This will create a lot of memory pressure on etcd.

### Goals

<!--
List the specific goals of the KEP. What is it trying to achieve? How will we
know that this has succeeded?
-->

> - APIServer is able to split large etcd range queries into smaller chunks

Reviewer 2: IMO this is more of a "how" we're trying to achieve the goal listed in the next line rather than "what". It doesn't sound like a goal to me in that sense.

- Reduce etcd memory consumption when kube clients list a large number of resources

### Non-Goals

<!--
What is out of scope for this KEP? Listing non-goals helps to focus discussion
and make progress.
-->

> - Throttle large range requests on etcd server

Reviewer 2: Can we generalize this as "Implementing any priority or fairness mechanisms for list calls made to etcd"?

> - Streaming list response from etcd to APIServer in chunks

Reviewer 2: "Streaming" and "chunks" seem to be a bit contradicting. We're actually proposing the latter in this KEP, but not the former. Let's make it clear.

> - Fine-tune with [priority and fairness](https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/1040-priority-and-fairness) or `max-requests-in-flight` to prevent etcd blowing up

Reviewer 2: Suggest -> "Reduce list call load on etcd using priority & fairness settings on the apiserver"

Author: Sounds good. I will update with `Reduce list call load on etcd using priority & fairness setting on apiserver`

Trace[638277422]: ---"Listing from storage done" 2268ms (23:00:00.844)
Trace[638277422]: [2.272115626s] [2.272115626s] END
```
10x clients issued such queries lead to etcd OOM-killed and failed other APIServer requests hitting to that node.

With the approach and `max-list-etcd-limit = 500`, user should expect 
- APIServer is healthy and serving requests because etcd cluster is healthy 
- List etcd duration increases from 2 seconds around 3 seconds and this duration is available as a prometheus metric exposed in "/metrics" endpoint. 
#### Story 2

### Notes/Constraints/Caveats (Optional) 

<!--
What are the caveats to the proposal?
What are some important details that didn't come across above?
Go in to as much detail as necessary here.
This might be a good place to talk about core concepts and how they relate.
-->

- Setting the maximum limit too low will increase the total round-trip time between APIServer and etcd to breach the default 1 minute request time. This will cause APIServer cancelles the requests prematurely. Example https://github.com/kubernetes/kubernetes/pull/94303. 

> - Use the maximum limit as suggestion and adjust it dynamically to fit in the 1 minute request timeout. For example, assuming each resource object size is identical given resource type, if the very first list with suggested maximum limit takes 2s but requires another 60 such subsequent list, double the limit to reduce the round trip time. 

Reviewer 2: Dynamically choosing the page size (that too per resource-type) is not just tedious, but also hard to get right IMO. For e.g these objects can grow over time, or round-trip duration may vary considerably in setups where etcd and apiserver are talking to each other over a network. Also the terminology "adjust it dynamically" isn't clear here - whether it is an operator externally changing the flag values on apiserver or apiserver internally adjusting its values. I'd suggest making these points clear in the caveat.

+cc @wojtek-t - for any thoughts

Reviewer 2: Also another perspective at this would be to push the client itself to paginate its list calls to apiserver, to avoid running into the above situation in the first place. This is also a general k8s guidance.

Author: After internal experiment and testing, I don't think dynamically choosing the page size is necessary as this only changes requests in a subset of list calls (almost every well-behaved client requests will be untouched and the behavior is just as it is today)

So, I'd like to remove this block.

How will UX be reviewed, and by whom?

Consider including folks who also work outside the SIG or subproject.
-->

## Design Details

<!--
This section should contain enough information that the specifics of your
change are understandable. This may include API specs (though not always
required) or even code snippets. If there's any ambiguity about HOW your
proposal will be implemented, this is the place to discuss them.
-->

Modify [apiserver/blob/master/pkg/storage/etcd3/store.go](https://github.com/kubernetes/apiserver/blob/master/pkg/storage/etcd3/store.go#L565-L576) list to apply a maximum limit to every etcd range request and issue multiple range requests if necessary to satisfy the original list. Filtering and decoding on each etcd response before issuing the next range request. Using the ResourceVersion that comes back from the first range request from etcd (or the one user provided) on later requests to ensure a consistent snapshot of resource collections. The implementation should unify with [current pagination implementation](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go#L686)

> The maximum limit will be a per-resource-type limit and exposed as an optional APIServer flag. The current behavior will not be changed by default.

Reviewer 2: change to -> `The new default behavior will continue to be the same as before when the flag is unspecified, i.e no max page-size will be imposed on the list calls to etcd`

for the KEP to move to `implementable` status and be included in the release.

In some cases, the questions below should also have answers in `kep.yaml`. This
is to enable automation to verify the presence of the review, and to reduce review
burden and latency.

The KEP must have a approver from the
[`prod-readiness-approvers`](http://git.k8s.io/enhancements/OWNERS_ALIASES)
team. Please reach out on the
[#prod-readiness](https://kubernetes.slack.com/archives/CPNHUMN74) channel if
you need any help or guidance.
-->

### Feature Enablement and Rollback

<!--
This section must be completed when targeting alpha to a release.
-->

###### How can this feature be enabled / disabled in a live cluster?

<!--
Pick one of these and delete the rest.

Documentation is available on [feature gate lifecycle] and expectations, as
well as the [existing list] of feature gates.

[feature gate lifecycle]: https://git.k8s.io/community/contributors/devel/sig-architecture/feature-gates.md
[existing list]: https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/
-->

- [ ] Feature gate (also fill in values in `kep.yaml`)
  - Feature gate name:

>   - Components depending on the feature gate:

Reviewer 2: Add feature gate name here and also `kube-apiserver` as the component.

- [ ] Other
  - Describe the mechanism:

>   - Will enabling / disabling the feature require downtime of the control

Reviewer 2: I believe this should be "Yes, this requires restart of the apiserver instance. However, there shouldn't be a downtime in HA setups where there are at least one replica kept active at any given time during the update."

+cc @liggitt - is there a proper wording for this already?
